libfreefare/mifare_desfire_key.c

```diff
@@ -124,16 +127,26 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
 void
 mifare_desfire_key_set_version(MifareDESFireKey key, uint8_t version)
 {
+    if (key->type == T_AES) {
+	key->aes_version = version;
+	return;
+    }
+
     for (int n = 0; n < 8; n++) {
 	uint8_t version_bit = ((version & (1 << (7 - n))) >> (7 - n));
 	key->data[n] &= 0xfe;
 	key->data[n] |= version_bit;
-	if (key->type == T_DES) {
+	switch (key->type) {
+	case T_DES:
 	    key->data[n + 8] = key->data[n];
-	} else {
+	    break;
+	case T_3DES:
 	    // Write ~version to avoid turning a 3DES key into a DES key
```

Comment:
"Now that the parity bits of key bites [0..7] are set to version, write ~version in parity bits of key bytes [8..15] in order to avoid 3DES keys to become DES keys at the PICC level."

Reply:
I've clarified the comment.